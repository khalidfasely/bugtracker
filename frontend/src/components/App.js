function App() {
  return (
    <div className="App">
      <header>
        Hello world from React        
      </header>
    </div>
  );
}

export default App;
