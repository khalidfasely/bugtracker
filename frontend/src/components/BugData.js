const BugData = ({ bug }) => (
    <div>
        <p>{bug?.id}</p>
        <p>{bug?.user?.username}</p>
        <p>{bug?.title}</p>
        <p>{bug?.description}</p>
        <p>{bug?.time}</p>
        <p>{bug?.classification}</p>
        <p>{`${bug?.active}`}</p>
        <div>
            {bug?.admins?.map(admin => <span key={admin} >{admin} </span>)}
        </div>
        <br />
        <div>
            {bug?.users_with?.map(user => <span key={user} >{user} </span>)}
        </div>
    </div>
);

export default BugData;