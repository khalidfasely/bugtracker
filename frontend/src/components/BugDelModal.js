import Modal from "react-modal";

const BugDelModal = ({ delModalOpen, setDelModalOpen, deleteBug }) => (
    <Modal
        isOpen={delModalOpen}
        onRequestClose={() => setDelModalOpen(false)}
    >
        If you delete this bug there is no way to return it!
        <button onClick={deleteBug} data-testid='delete_button'>Delete</button>
    </Modal>
);

export default BugDelModal;