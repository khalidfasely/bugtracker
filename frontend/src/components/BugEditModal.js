import Modal from "react-modal";

import NewBug from "./NewBug";

const BugEditModal = ({
    editModalOpen, setEditModalOpen,
    users, bug
}) => (
    <Modal
        isOpen={editModalOpen}
        onRequestClose={() => setEditModalOpen(false)}
    >
        <NewBug users={users} isEdit={true} bug={bug} setEditModalOpen={setEditModalOpen} />
    </Modal>
);

export default BugEditModal;