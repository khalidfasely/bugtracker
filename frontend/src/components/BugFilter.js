import { connect } from "react-redux";
import { setTextFilter, sortByDefault, sortByPriority } from "../actions/bugFilter";

export const BugFilter = ({ filters, setTextFilter, sortByDefault, sortByPriority }) => {
    const onSortByChange = (e) => {
        if (e.target.value === 'priority') {
            sortByPriority();
            return;
        };
        sortByDefault();
    };

    return (
        <div className="bugs-filter">
            <div className="input-group__item">
                <input
                    data-testid='search_input'
                    type="text"
                    className="text-input"
                    placeholder="Search bugs"
                    value={filters.text}
                    maxLength={35}
                    onChange={(e) => setTextFilter(e.target.value)}
                />
            </div>
            <div className="input-group__item">
                <select
                    data-testid='select_input'
                    className="select"
                    value={filters.sortBy}
                    onChange={onSortByChange}
                >
                    <option value="">Date</option>
                    <option value="priority">Priority</option>
                </select>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => ({
    filters: state.bugFilter
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByPriority: () => dispatch(sortByPriority()),
    sortByDefault: () => dispatch(sortByDefault())
});

export default connect(mapStateToProps, mapDispatchToProps)(BugFilter);