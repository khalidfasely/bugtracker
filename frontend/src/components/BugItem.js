import { Link } from "react-router-dom";
import { connect } from "react-redux";

import Comment from "./Comment";

const BugItem = ({uname, bug, comments}) => {
    console.log(bug, comments);
    return (
        <>
            {
                bug?.users_with?.includes(uname) ?
                <div>
                    **** bug <br /><h1>{bug.title}</h1>
                    **** comments {comments?.map(comment => <Comment comment={comment} />)}
                </div> :
                <div>Not able to see this bug! <Link to="/new">Create one!</Link></div>
            }
        </>
    );
}

const mapStateToProps = (state) => ({
    uname: state.auth.uname,
    bug: state.bugItem.bug,
    comments: state.bugItem.comments
});

export default connect(mapStateToProps)(BugItem);