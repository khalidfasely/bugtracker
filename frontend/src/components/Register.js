import { useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { connect } from 'react-redux';
import { startRegister } from '../actions/auth';
import { startSetProjects } from '../actions/projects';
import Loading from './Loading';

export const Register = ({ startRegister, startSetProjects }) => {
    const history = useNavigate();
    const [loading, setLoading] = useState(false);
    const [username, setUsername] = useState('');
    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');
    const [shownPd, setShownPd] = useState(false);
    const [confirmation, setConfirmation] = useState('');
    const [shownCfr, setShownCfr] = useState(false);
    const [error, setError] = useState('');

    const onFormSubmit = async (e) => {
        e.preventDefault();
        setError('');

        const availabeValues = username && email && password && confirmation;

        // Check all fields
        if (!availabeValues) {
            setError('Must set all fields!');
            return;
        };

        // Compare passwords
        if (password !== confirmation) {
            setError('Password and Confirmation must match!');
            return;
        };

        setLoading(true);

        // Start register with redux
        const result = await startRegister({
            username,
            email,
            password,
            confirmation
        });

        // Check result from backend
        if (result.message !== "Register correctly.") {
            setLoading(false);
            setError(result.message);
        } else {
            setUsername('');
            setEmail('');
            setPassword('');
            setConfirmation('');

            // Reset the projects after register and login
            startSetProjects();

            history('/');
        };
    };

    if (loading) {
        return (
            <div className='register-container'>
                <div className='register'>
                    <Loading />
                </div>
            </div>
        );
    }

    return (
        <div className='register-container'>
            <div className='register'>
                <form onSubmit={onFormSubmit} className='register-form' data-testid='register_form'>
                    <p className='error_message' data-testid='register_error'>{error ? error : null}</p>
                    <input
                        type='text'
                        autoFocus
                        placeholder='Username'
                        value={username}
                        onChange={(e) => setUsername(e.target.value)}
                        data-testid='username_input'
                    />
                    <input
                        type='email'
                        placeholder='Email'
                        value={email}
                        onChange={(e) => setEmail(e.target.value)}
                        data-testid='email_input'
                    />
                    <input
                        type={shownPd ? 'text' : 'password'}
                        placeholder='Password'
                        value={password}
                        onChange={(e) => setPassword(e.target.value)}
                        data-testid='password_input'
                    />
                    {
                        shownPd ?
                        <button className='show-pswrd' type="button" onClick={() => setShownPd(false)}>Hide password!</button> :
                        <button className='show-pswrd' disabled={!password} type="button" onClick={() => setShownPd(true)}>Show password!</button>
                    }
                    <input
                        type={shownCfr ? 'text' : 'password'}
                        placeholder='Confirmation'
                        value={confirmation}
                        onChange={(e) => setConfirmation(e.target.value)}
                        data-testid='confirmation_input'
                    />
                    {
                        shownCfr ?
                        <button className='show-pswrd' type="button" onClick={() => setShownCfr(false)}>Hide confirmation!</button> :
                        <button className='show-pswrd' disabled={!confirmation} type="button" onClick={() => setShownCfr(true)}>Show confirmation!</button>
                    }
                    <button className='register__button' data-testid='register_button'>Register</button>
                </form>
                <p>Already have an account? Login in <Link to='/login'>here</Link>.</p>
            </div>
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startRegister: (newUserData) => dispatch(startRegister(newUserData)),
    startSetProjects: () => dispatch(startSetProjects())
});

export default connect(undefined, mapDispatchToProps)(Register);