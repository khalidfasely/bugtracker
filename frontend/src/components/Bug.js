import { useState, useEffect } from "react";
import { connect } from "react-redux";
import { useParams } from "react-router-dom";
import { startSetBug } from "../actions/bug";

import BugItem from './BugItem';

export const Bug = ({ startSetBug }) => {
    const params = useParams();

    const [ableToSee, setAbleToSee] = useState(false);
    const [notFound, setNotFound] = useState('');

    useEffect(() => {
        startSetBug(params.pid, params.bid)
        .then((result) => {
            setNotFound('');

            if (result.message) {
                setNotFound(result.message);
            };
        })
        .then(() => setAbleToSee(true))
        .catch(er => console.error(er));
    }, [params])


    if (notFound) {
        return <div>Bug Not Found! -- {notFound}</div>;
    };

    return (
        <div>
            {
                ableToSee ?
                <div><BugItem /></div> :
                <div>You're not able to see this bug</div>
            }
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startSetBug: (pid, bid) => dispatch(startSetBug(pid, bid))
});

export default connect(undefined, mapDispatchToProps)(Bug);