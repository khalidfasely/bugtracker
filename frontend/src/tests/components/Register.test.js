import React from 'react';
import { Register } from '../../components/Register';
import { render, fireEvent, act } from '@testing-library/react';

test('Should render Register component correctly', () => {
    const { asFragment } = render(<Register />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render button with correct text', () => {
    const { getByTestId } = render(<Register />);
    const buttonEl = getByTestId('register_button');
    expect(buttonEl.textContent).toBe('Register');
});

test('Should change username input value', () => {
    const { getByTestId } = render(<Register />);
    const usernameInputEl = getByTestId('username_input');

    expect(usernameInputEl.value).toBe('');

    fireEvent.change(usernameInputEl, {
        target: {
            value: 'Username'
        }
    });

    expect(usernameInputEl.value).toBe('Username');
});

test('Should change email input value', () => {
    const { getByTestId } = render(<Register />);
    const emailInputEl = getByTestId('email_input');

    expect(emailInputEl.value).toBe('');

    fireEvent.change(emailInputEl, {
        target: {
            value: 'Email@test'
        }
    });

    expect(emailInputEl.value).toBe('Email@test');
});

test('Should change password input value', () => {
    const { getByTestId } = render(<Register />);
    const passwordInputEl = getByTestId('password_input');

    expect(passwordInputEl.value).toBe('');

    fireEvent.change(passwordInputEl, {
        target: {
            value: 'Password'
        }
    });

    expect(passwordInputEl.value).toBe('Password');
});

test('Should change confirmation input value', () => {
    const { getByTestId } = render(<Register />);
    const confirmationInputEl = getByTestId('confirmation_input');

    expect(confirmationInputEl.value).toBe('');

    fireEvent.change(confirmationInputEl, {
        target: {
            value: 'Confirmation'
        }
    });

    expect(confirmationInputEl.value).toBe('Confirmation');
});

test('Should not submit the form with empty inputs', () => {
    const { getByTestId } = render(<Register />);
    const formEl = getByTestId('register_form');

    fireEvent.submit(formEl);

    const errorEl = getByTestId('register_error');

    expect(errorEl.textContent).toBe('Must set all fields!');
});

test('Should not submit the form with one empty input', () => {
    const { getByTestId } = render(<Register />);
    const formEl = getByTestId('register_form');

    const usernameInputEl = getByTestId('username_input');
    const emailInputEl = getByTestId('email_input');
    const passwordInputEl = getByTestId('password_input');

    fireEvent.change(usernameInputEl, {
        target: {
            value: 'Username'
        }
    });

    fireEvent.change(emailInputEl, {
        target: {
            value: 'Email@test'
        }
    });

    fireEvent.change(passwordInputEl, {
        target: {
            value: 'Password'
        }
    });

    fireEvent.submit(formEl);

    const errorEl = getByTestId('register_error');

    expect(errorEl.textContent).toBe('Must set all fields!');
});

test('Should not submit the form if password and confirmation does not matchs', () => {
    const { getByTestId } = render(<Register />);
    const formEl = getByTestId('register_form');

    const usernameInputEl = getByTestId('username_input');
    const emailInputEl = getByTestId('email_input');
    const passwordInputEl = getByTestId('password_input');
    const confirmationInputEl = getByTestId('confirmation_input');

    fireEvent.change(usernameInputEl, {
        target: {
            value: 'Username'
        }
    });

    fireEvent.change(emailInputEl, {
        target: {
            value: 'Email@test'
        }
    });

    fireEvent.change(passwordInputEl, {
        target: {
            value: 'Password'
        }
    });

    fireEvent.change(confirmationInputEl, {
        target: {
            value: 'AnotherPassword'
        }
    });

    fireEvent.submit(formEl);

    const errorEl = getByTestId('register_error');

    expect(errorEl.textContent).toBe('Password and Confirmation must match!');
});

test('Should submit form with full inputs', async () => {
    const promise = Promise.resolve({});
    const startRegister = jest.fn(() => promise);

    const { getByTestId } = render(<Register startRegister={startRegister} />);
    const formEl = getByTestId("register_form");

    const usernameInputEl = getByTestId("username_input");
    fireEvent.change(usernameInputEl, {
        target: {
            value: 'Username'
        }
    });

    const emailInputEl = getByTestId("email_input");
    fireEvent.change(emailInputEl, {
        target: {
            value: 'Email@o.com'
        }
    });

    const passwordInputEl = getByTestId("password_input");
    fireEvent.change(passwordInputEl, {
        target: {
            value: 'Password'
        }
    });

    const confirmationInputEl = getByTestId("confirmation_input");
    fireEvent.change(confirmationInputEl, {
        target: {
            value: 'Password'
        }
    });

    expect(startRegister).toHaveBeenCalledTimes(0);

    fireEvent.submit(formEl, {
        preventDefault: () => {}
    });

    await act(() => promise);
    expect(startRegister).toHaveBeenCalledTimes(1);
});