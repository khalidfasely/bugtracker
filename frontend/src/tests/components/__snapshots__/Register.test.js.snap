// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Register component correctly 1`] = `
<DocumentFragment>
  <div
    class="register-container"
  >
    <div
      class="register"
    >
      <form
        class="register-form"
        data-testid="register_form"
      >
        <p
          class="error_message"
          data-testid="register_error"
        />
        <input
          data-testid="username_input"
          placeholder="Username"
          type="text"
          value=""
        />
        <input
          data-testid="email_input"
          placeholder="Email"
          type="email"
          value=""
        />
        <input
          data-testid="password_input"
          placeholder="Password"
          type="password"
          value=""
        />
        <button
          class="show-pswrd"
          disabled=""
          type="button"
        >
          Show password!
        </button>
        <input
          data-testid="confirmation_input"
          placeholder="Confirmation"
          type="password"
          value=""
        />
        <button
          class="show-pswrd"
          disabled=""
          type="button"
        >
          Show confirmation!
        </button>
        <button
          class="register__button"
          data-testid="register_button"
        >
          Register
        </button>
      </form>
      <p>
        Already have an account? Login in 
        <a>
          here
        </a>
        .
      </p>
    </div>
  </div>
</DocumentFragment>
`;
